File: muse/muse/Lyrics/JLF.swift
import Foundation

// MARK: - JLF (JSON Lyrics Format)

struct JLF: Codable, Sendable {
    var lines: SyncedLines
    var richsync: SyncedRich?
    var metadata: SyncedMetadata?
    var source: String
    var name: String?
    var message: String?
}

struct SyncedMetadata: Codable, Sendable {
    var mxmId: String?
    var iTunesId: String?
    var spotifyId: String?
    var artist: String
    var title: String
    var album: String
    var copyright: String?

    enum CodingKeys: String, CodingKey {
        case mxmId = "MxmId"
        case iTunesId = "ITunesId"
        case spotifyId = "SpotifyId"
        case artist = "Artist"
        case title = "Title"
        case album = "Album"
        case copyright = "Copyright"
    }
}

struct SyncedLines: Codable, Sendable {
    var lines: [SyncedLine]
    var linesEnd: Int

    init(lines: [SyncedLine], linesEnd: Int) {
        self.lines = lines
        self.linesEnd = linesEnd
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        lines = try container.decode([SyncedLine].self, forKey: .lines)
        if let d = try? container.decode(Double.self, forKey: .linesEnd) {
            linesEnd = Int(d * 1000)
        } else {
            let s = try container.decode(String.self, forKey: .linesEnd)
            linesEnd = Int((Double(s) ?? 0) * 1000)
        }
    }
}

struct SyncedLine: Codable, Sendable, Identifiable {
    var time: Int  // ms since start of song
    var text: String
    var translation: String?

    var id: Int { time }

    init(time: Int, text: String, translation: String? = nil) {
        self.time = time
        self.text = text
        self.translation = translation
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        time = Int((try container.decode(Double.self, forKey: .time)) * 1000)
        text = try container.decodeIfPresent(String.self, forKey: .text) ?? ""
        translation = try container.decodeIfPresent(String.self, forKey: .translation)
    }
}

// MARK: - Rich Sync (for future use)

struct SyncedRich: Codable, Sendable {
    var totalTime: Int
    var sections: [SyncedRichSection]
    var agents: [SyncedRichAgent]

    init(totalTime: Int, sections: [SyncedRichSection], agents: [SyncedRichAgent]) {
        self.totalTime = totalTime
        self.sections = sections
        self.agents = agents
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        totalTime = Int((try container.decode(Double.self, forKey: .totalTime)) * 1000)
        sections = try container.decode([SyncedRichSection].self, forKey: .sections)
        agents = try container.decode([SyncedRichAgent].self, forKey: .agents)
    }
}

enum SyncedRichAgentType: String, Codable, Sendable {
    case person
    case group
    case other
}

struct SyncedRichAgent: Codable, Sendable {
    var type: String
    var id: String
}

struct SyncedRichSection: Codable, Sendable {
    var timeStart: Int
    var timeEnd: Int
    var lines: [SyncedRichLine]

    init(timeStart: Int, timeEnd: Int, lines: [SyncedRichLine]) {
        self.timeStart = timeStart
        self.timeEnd = timeEnd
        self.lines = lines
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        timeStart = Int((try container.decode(Double.self, forKey: .timeStart)) * 1000)
        timeEnd = Int((try container.decode(Double.self, forKey: .timeEnd)) * 1000)
        lines = try container.decode([SyncedRichLine].self, forKey: .lines)
    }
}

struct SyncedRichLine: Codable, Sendable {
    var timeStart: Int
    var timeEnd: Int
    var text: String
    var segments: [SyncedRichLineSegment]
    var agent: String
    var bgVox: SyncedRichBackgroundLine?

    init(timeStart: Int, timeEnd: Int, text: String, segments: [SyncedRichLineSegment], agent: String, bgVox: SyncedRichBackgroundLine? = nil) {
        self.timeStart = timeStart
        self.timeEnd = timeEnd
        self.text = text
        self.segments = segments
        self.agent = agent
        self.bgVox = bgVox
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        timeStart = Int((try container.decode(Double.self, forKey: .timeStart)) * 1000)
        timeEnd = Int((try container.decode(Double.self, forKey: .timeEnd)) * 1000)
        text = try container.decode(String.self, forKey: .text)
        segments = try container.decode([SyncedRichLineSegment].self, forKey: .segments)
        agent = try container.decode(String.self, forKey: .agent)
        bgVox = try container.decodeIfPresent(SyncedRichBackgroundLine.self, forKey: .bgVox)
    }
}

struct SyncedRichBackgroundLine: Codable, Sendable {
    var timeStart: Int
    var timeEnd: Int
    var text: String
    var segments: [SyncedRichLineSegment]

    init(timeStart: Int, timeEnd: Int, text: String, segments: [SyncedRichLineSegment]) {
        self.timeStart = timeStart
        self.timeEnd = timeEnd
        self.text = text
        self.segments = segments
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        timeStart = Int((try container.decode(Double.self, forKey: .timeStart)) * 1000)
        timeEnd = Int((try container.decode(Double.self, forKey: .timeEnd)) * 1000)
        text = try container.decode(String.self, forKey: .text)
        segments = try container.decode([SyncedRichLineSegment].self, forKey: .segments)
    }
}

struct SyncedRichLineSegment: Codable, Sendable {
    var text: String
    var timeStart: Int
    var timeEnd: Int

    init(text: String, timeStart: Int, timeEnd: Int) {
        self.text = text
        self.timeStart = timeStart
        self.timeEnd = timeEnd
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        text = try container.decode(String.self, forKey: .text)
        timeStart = Int((try container.decode(Double.self, forKey: .timeStart)) * 1000)
        timeEnd = Int((try container.decode(Double.self, forKey: .timeEnd)) * 1000)
    }
}

File: muse-web/src/lyrics/jlf.ts
/**
 * JLF (JSON Lyrics Format) — the response shape from umi's `/lyrics` endpoint.
 *
 * Ported from the iOS (`muse/muse/Lyrics/JLF.swift`) and Android
 * (`muse-android/.../lyrics/Jlf.kt`) models. umi returns all timestamps as
 * **seconds** (floating point or string). We convert to **milliseconds**
 * internally so the lyrics view can compare against `player.currentTime * 1000`.
 */

// --------------------------------------------------------------- raw JSON shape

/** Raw JLF as it arrives from umi — times are in seconds. */
export interface RawJlf {
  lines: RawSyncedLines;
  richsync?: RawSyncedRich | null;
  metadata?: RawSyncedMetadata | null;
  source: string;
  name?: string | null;
  message?: string | null;
}

export interface RawSyncedLines {
  lines: RawSyncedLine[];
  /** seconds — number or string (umi is inconsistent). */
  linesEnd: number | string;
}

export interface RawSyncedLine {
  /** seconds */
  time: number | string;
  text?: string;
  translation?: string | null;
}

export interface RawSyncedMetadata {
  MxmId?: string;
  ITunesId?: string;
  SpotifyId?: string;
  Artist: string;
  Title: string;
  Album: string;
  Copyright?: string;
}

export interface RawSyncedRich {
  /** seconds */
  totalTime: number | string;
  sections: RawSyncedRichSection[];
  agents?: RawSyncedRichAgent[];
}

export interface RawSyncedRichAgent {
  type: string;
  id: string;
}

export interface RawSyncedRichSection {
  /** seconds */
  timeStart: number | string;
  timeEnd: number | string;
  lines: RawSyncedRichLine[];
}

export interface RawSyncedRichLine {
  /** seconds */
  timeStart: number | string;
  timeEnd: number | string;
  text: string;
  segments: RawSyncedRichLineSegment[];
  agent: string;
  bgVox?: RawSyncedRichBackgroundLine | null;
}

export interface RawSyncedRichBackgroundLine {
  /** seconds */
  timeStart: number | string;
  timeEnd: number | string;
  text: string;
  segments: RawSyncedRichLineSegment[];
}

export interface RawSyncedRichLineSegment {
  text: string;
  /** seconds */
  timeStart: number | string;
  timeEnd: number | string;
}

// --------------------------------------------------------------- parsed (ms)

/** Parsed JLF — all times converted to integer milliseconds. */
export interface Jlf {
  lines: SyncedLines;
  richsync: SyncedRich | null;
  metadata: SyncedMetadata | null;
  source: string;
  name: string | null;
  message: string | null;
}

export interface SyncedLines {
  lines: SyncedLine[];
  linesEnd: number;
}

export interface SyncedLine {
  /** ms since song start */
  time: number;
  text: string;
  translation: string | null;
  /** stable id for React keys */
  id: number;
}

export interface SyncedMetadata {
  mxmId: string | null;
  iTunesId: string | null;
  spotifyId: string | null;
  artist: string;
  title: string;
  album: string;
  copyright: string | null;
}

export interface SyncedRich {
  totalTime: number;
  sections: SyncedRichSection[];
  agents: SyncedRichAgent[];
}

export interface SyncedRichAgent {
  type: string;
  id: string;
}

export interface SyncedRichSection {
  timeStart: number;
  timeEnd: number;
  lines: SyncedRichLine[];
}

export interface SyncedRichLine {
  timeStart: number;
  timeEnd: number;
  text: string;
  segments: SyncedRichLineSegment[];
  agent: string;
  bgVox: SyncedRichBackgroundLine | null;
}

export interface SyncedRichBackgroundLine {
  timeStart: number;
  timeEnd: number;
  text: string;
  segments: SyncedRichLineSegment[];
}

export interface SyncedRichLineSegment {
  text: string;
  timeStart: number;
  timeEnd: number;
}

// --------------------------------------------------------------- conversion

function toMs(v: number | string | undefined | null): number {
  if (v == null) return 0;
  const n = typeof v === "string" ? parseFloat(v) : v;
  return Math.round((n || 0) * 1000);
}

export function parseJlf(raw: RawJlf): Jlf {
  return {
    lines: {
      lines: (raw.lines?.lines ?? []).map((l, i) => ({
        time: toMs(l.time),
        text: l.text ?? "",
        translation: l.translation ?? null,
        id: i,
      })),
      linesEnd: toMs(raw.lines?.linesEnd),
    },
    richsync: raw.richsync
      ? {
          totalTime: toMs(raw.richsync.totalTime),
          sections: (raw.richsync.sections ?? []).map((s) => ({
            timeStart: toMs(s.timeStart),
            timeEnd: toMs(s.timeEnd),
            lines: (s.lines ?? []).map((l) => ({
              timeStart: toMs(l.timeStart),
              timeEnd: toMs(l.timeEnd),
              text: l.text ?? "",
              segments: (l.segments ?? []).map((seg) => ({
                text: seg.text,
                timeStart: toMs(seg.timeStart),
                timeEnd: toMs(seg.timeEnd),
              })),
              agent: l.agent ?? "",
              bgVox: l.bgVox
                ? {
                    timeStart: toMs(l.bgVox.timeStart),
                    timeEnd: toMs(l.bgVox.timeEnd),
                    text: l.bgVox.text ?? "",
                    segments: (l.bgVox.segments ?? []).map((seg) => ({
                      text: seg.text,
                      timeStart: toMs(seg.timeStart),
                      timeEnd: toMs(seg.timeEnd),
                    })),
                  }
                : null,
            })),
          })),
          agents: raw.richsync.agents ?? [],
        }
      : null,
    metadata: raw.metadata
      ? {
          mxmId: raw.metadata.MxmId ?? null,
          iTunesId: raw.metadata.ITunesId ?? null,
          spotifyId: raw.metadata.SpotifyId ?? null,
          artist: raw.metadata.Artist,
          title: raw.metadata.Title,
          album: raw.metadata.Album,
          copyright: raw.metadata.Copyright ?? null,
        }
      : null,
    source: raw.source,
    name: raw.name ?? null,
    message: raw.message ?? null,
  };
}
